3155 prioritized errors

elipe17 opened the pull request:
## Summary of Changes
- Added tech memo describing findings on current OFA requirements for prioritizing errors
- Implemented tech memo
  - Updated XLSX generation to use pagination to avoid bringing entire queryset into memory
  - Added queryset filtering based on provided "priority" errors in [this OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=2ZrCcZ).
  ~~- Updated file status to be partially accepted with errors if more than 500 priority errors are detected~~
  - Updated spreadsheet generator to have a "Critical" and "Summary" tab. Critical is all priority errors and the Summary tab contains all errors grouped by their error message, year, month, and item names/numbers
  - Updated logic so that error report download is user group agnostic

Pull request closes #3155 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Unzip and submit the attached file `WA_SSPActive_anon.txt`
3. Confirm the status is Partially Accepted with Errors
4. Download the report and verify you get the two tabs discussed above

[WA_SSPActive_anon.txt.zip](https://github.com/user-attachments/files/17464255/WA_SSPActive_anon.txt.zip)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Tech memo outlining design/method to prioritize errors for reports
+ [x] Memory pressure relieved during report generation no matter the size of the queryset

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Update xlsx generator to pin memory consumption via paginator

elipe17 committed:
- fix white space

elipe17 committed:
- Add logic to narrow error report when stt is accessing it
- Add admin property to user model
- whitespace clean up

elipe17 committed:
- Add explainer

elipe17 committed:
- add final query

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3155-prioritized-errors

elipe17 committed:
- add logs to ignore

elipe17 committed:
- Updated queries
- Ordered based on OneNote

elipe17 committed:
- Added tech memo outlining prioritized errors work

elipe17 committed:
- fix tests

elipe17 committed:
- Fixed None string included in report

elipe17 committed:
- Change queryset ordering

elipe17 committed:
- Adding debug code

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3155-prioritized-errors

elipe17 committed:
- enable layer caching

elipe17 committed:
- use same method to get worksheet throughout tests
- delete workbook if it exists before creating a new one
- Remove debug code

elipe17 committed:
- Update error messages

elipe17 committed:
- move order_by to be the last query
- Update row error message number

elipe17 committed:
- Update order by clause
- add debug code

elipe17 committed:
- remove debug code

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

elipe17 committed:
- Assign order by clause

jtimpe commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+Given the current OFA requirements, we can implement prioritized errors, and memory efficient report generation without too much work. OFA has provided [this OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) document which outlines the error types, errors, and fields that are most important/prioritized for STTs to see.
+
+### Memory Efficient Report Generation
+As previously mentioned in the #background section, the `get_xls_serialized_file` introduces a method to serialize parser errors into a XLSX that requires the entire queryset of parser errors to be brought into memory. Because these querysets can be very large, having them in memory regularly kills Celery workers with an OOM error. To remedy the issue, this tech memo suggests updating `get_xls_se ...
```
`regularly kills Celery workers with an OOM error`

i think the Gunicorn worker would be what's getting killed, since the parser error report generation is not done in a celery task

jtimpe commented on tdrs-backend/tdpservice/data_files/util.py:
```diff
@@ -4,54 +4,40 @@
+
+def friendly_names(fields_json):
+    """Return comma separated string of friendly names."""
+    return ','.join([i for i in fields_json['friendly_name'].values()])
```
maybe consider adding `if 'friendly_name` in fields_json` or something to that effect, just in case the errors json doesn't have anything in it.

jtimpe replied:
never mind, i see that this is done in `check_fields_json` below

jtimpe commented on tdrs-backend/tdpservice/data_files/util.py:
```diff
@@ -91,16 +77,36 @@ def format_header(header_list: list):
+            worksheet.write(row_idx, 5, friendly_names(fields_json))
+            worksheet.write(row_idx, 6, internal_names(fields_json))
+            worksheet.write(row_idx, 7, record.row_number)
+            worksheet.write(row_idx, 8, str(ParserErrorCategoryChoices(record.error_type).label))
```
much cleaner!

jtimpe commented on tdrs-backend/tdpservice/data_files/views.py:
```diff
@@ -147,9 +147,89 @@ def download(self, request, pk=None):
+                    "CITIZENSHIP_STATUS",
+                    "WORK_ELIGIBLE_INDICATOR",
+                    ],
+                    error_type=ParserErrorCategoryChoices.VALUE_CONSISTENCY))
```
is there a way to shorten this function? perhaps a function that does `filtered_errors.union` and takes a list param for `fields_json__friendly_name__has_keys` and an `error_type` param. or different functions for different user roles, for `is_active` and `is_closed` etc. 

just a lot of lines here, figured we could reduce the cognitive complexity somehow.

jtimpe reviewed (approved):
Works as designed. Couple minor suggestions for improvement but makes the error report significantly smaller and faster to download. Good work!

raftmsohani commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+# TDP Prioritized Parser Errors
```
Like the documentation 🥇

raftmsohani commented on tdrs-backend/tdpservice/data_files/views.py:
```diff
@@ -147,9 +147,89 @@ def download(self, request, pk=None):
+        is_closed = "Closed" in datafile.section
+
+        # We only filter Active and Closed submissions. Aggregate and Stratum return all errors.
+        if not (user.is_ofa_sys_admin or user.is_ofa_admin) and (is_active or is_closed):
```
not directly related to your PR, but If we are intending to use `is_ofa_sys_admin` or `user.is_ofa_admin` together, it is better to cache these properties since these will be two queries

andrew-jameson commented on .circleci/base_config.yml:
```diff
@@ -12,11 +12,11 @@ executors:
   machine-executor:
     machine:
-      docker_layer_caching: false
+      docker_layer_caching: true
```
Is this for Nexus or something else? A little confused on why this is in this PR

elipe17 replied:
I was just checking to see if it had an effect on the build and test times. It doesn't unfortunately and I forgot to remove it.

elipe17 committed:
- Correct document
- functionize prioritized queries
- Update group check function to allow multiple groups
- add new admin property to user class

elipe17 committed:
- revert layer cachhing

raftmsohani reviewed (approved):
LGTM!

andrew-jameson commented on tdrs-backend/tdpservice/data_files/views.py:
```diff
@@ -143,13 +158,44 @@ def download(self, request, pk=None):
+                field_name__in=fields,
+                error_type=ParserErrorCategoryChoices.FIELD_VALUE
+            ))
+
```
These are so much more concise, great job!

ADPennington committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington commented:
very interesting work here @elipe17 😄 @klinkoberstar @ttran-hub -- lets plan to review this together next week too!

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington commented:
@elipe17 this is in good shape, but we have concerns about the submission history detail for STT Data Analysts:

- it still captures information about errors detected even when the prioritized error reports don't include any errors 
- the team is also concerned about the burden of having to share the full scope of errors detected to resolve the confusion introduced by the submission history detail. 
- @klinkoberstar will share some recommendations/questions about how we could help mitigate confusion, including adding a worksheet of prioritized errors vs suppressing de-prioritized errors.

klinkoberstar commented:
> @elipe17 this is in good shape, but we have concerns about the submission history detail for STT Data Analysts:
> 
> * it still captures information about errors detected even when the prioritized error reports don't include any errors
> * the team is also concerned about the burden of having to share the full scope of errors detected to resolve the confusion introduced by the submission history detail.
> * @klinkoberstar will share some recommendations/questions about how we could help mitigate confusion, including adding a worksheet of prioritized errors vs suppressing de-prioritized errors.

Jumping in here to add a few thoughts:

One suggestion is to include all the errors in the report, but have separate tabs:
a. First tab with the "High Priority Errors" (wording TBD) - the prioritized view
b. Second tab with "All Errors" - admin (full) view

This still achieves the goal of highlighting priority errors for STTs but also still gives access to the full report if they want it. Hopefully this prevents STTs from emailing OFA to get more details on these lower priority errors. My understanding is that States used to be able to filter warnings in TDRS with the edit indicator - so hopefully this wouldn't be too big of a leap.

One concern with this is the ballooning file size. Some submissions have a large number of errors which might make the file harder to download. One suggestion is to suppress the second report for submissions that have above a certain threshold of high priority errors. In this case, the second tab could just have a warning with something to the effect of "Due to the large volume of high priority errors, TDP is not able to generate the full error report. Please address the high priority errors in Tab 1 and resubmit to see the full error report." 

We'd probably want to consider some adjustments to the Submission Email if we head in this direction. We can fold this into #3251

ADPennington reviewed (changes requested):
Below is a summary of notes from the sync between @elipe17 @klinkoberstar  and me cc: @lfrohlich @ttran-hub:

in QASP review we noticed a couple of potential issues with the prioritization strategy:
- the submission history metadata is unaffected by the content included in prioritized error report (e.g. submission history could include errors and the report does not include any errors). This could yield confusion
- if the remaining errors are only available to the data team, STTs would need to reach out to the team for the full error report (burden for STTs and OFA)
- OFA would still have access to the complete version of the error report, but may still encounter issues downloading due to file size issues ([see #3026](https://github.com/raft-tech/TANF-app/issues/3026))

Given the above, we discussed a solution that would satisfy the following goals: 

1. giving STTs full visibility into errors detected by the system
2. flagging the most important errors for STTs to review
3. limiting the content included in the error report 
4. configuring a new email notification related to goal # 3. 

Recommendations:
- error report should include 2 tabs:
   - first tab includes _most important errors_
   - 2nd tab includes aggregate version of all errors detected by the system (grouped by: year, month, error message, item number(s), and item name(s)

- implement threshold: stop processing file after 500 of the _most important errors_ were detected. 
- error report content will be consistent across permissions groups with access to it. 
- send a unique e-mail notification when file status is "Rejected" or "Partially Accepted with errors" indicating that the file processing was interrupted and errors need to be reviewed and corrected (specific guidance to be scoped as part of [#3251](https://github.com/raft-tech/TANF-app/issues/3251))

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

elipe17 committed:
- Update report to include aggregates tab along with prioritized tab
- functionized util.py
- updated views.py to pass both querysets

elipe17 committed:
- linting

elipe17 committed:
- Added tests for aggregates page

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

elipe17 committed:
- linting

elipe17 committed:
Merge branch '3155-prioritized-errors' of https://github.com/raft-tech/TANF-app into 3155-prioritized-errors

elipe17 commented:
@ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?

ADPennington commented:
> @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?

i believe its the former: stop processing altogether when 500 of the most important errors were detected; status would be rejected
and error report would include the 500 most important errors detected 

our goal is for the STT to review these errors, correct them, and resubmit the file. its likely that this will be an iterative process for the STT since a file with 500 critical errors likely has a lot more than 500 critical errors. the email notification in #3251  will help set the appropriate expectations. 

and in case its worth noting for documentation purposes, this strategy is aligned with the legacy system behavior when 500 critical errors were detected (see [system edit code T1-501](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_system_edits_fatal_and_warning.pdf)). the system would return T1-501 as an error but did not include any information about what to fix, so we think this is a step in the right direction, in terms of providing actionable feedback while also limiting the load on the system in continuing to process an error-prone file. 

cc: @klinkoberstar @victoriaatraft @reitermb

klinkoberstar commented:
> @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?



> > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> 
> i believe its the former: stop processing altogether when 500 of the most important errors were detected; status would be rejected and error report would include the 500 most important errors detected
> 
> our goal is for the STT to review these errors, correct them, and resubmit the file. its likely that this will be an iterative process for the STT since a file with 500 critical errors likely has a lot more than 500 critical errors. the email notification in #3251 will help set the appropriate expectations.
> 
> and in case its worth noting for documentation purposes, this strategy is aligned with the legacy system behavior when 500 critical errors were detected (see [system edit code T1-501](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_system_edits_fatal_and_warning.pdf)). the system would return T1-501 as an error but did not include any information about what to fix, so we think this is a step in the right direction, in terms of providing actionable feedback while also limiting the load on the system in continuing to process an error-prone file.
> 
> cc: @klinkoberstar @victoriaatraft @reitermb

potential thought to consider re: how this decision will impact the submission history table. if we stop parsing all together at 500 high priority errors, then when STTs make changes to their files and resubmit to address these errors, it's theoretically possible that the total number of errors on the submission history page goes up on resubmission, not down (due to lower priority errors being evaluated that weren't originally). This may be counterintuitive. if we parse all the records but only show the first 500, the submission history page can still list the true # of records with errors - we will just limit how much detail we show them in the report. 

cc: @ADPennington

ADPennington commented:
Would serving up submission history metadata that is not aligned with the content of the error report be more or less confusing than receiving more/same number of errors after resubmitting a corrected file? 

no doubt that either way we go here we will need to help guide expectations 😊 @klinkoberstar

elipe17 commented:
> > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> 
> > > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> > 
> > 
> > i believe its the former: stop processing altogether when 500 of the most important errors were detected; status would be rejected and error report would include the 500 most important errors detected
> > our goal is for the STT to review these errors, correct them, and resubmit the file. its likely that this will be an iterative process for the STT since a file with 500 critical errors likely has a lot more than 500 critical errors. the email notification in #3251 will help set the appropriate expectations.
> > and in case its worth noting for documentation purposes, this strategy is aligned with the legacy system behavior when 500 critical errors were detected (see [system edit code T1-501](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_system_edits_fatal_and_warning.pdf)). the system would return T1-501 as an error but did not include any information about what to fix, so we think this is a step in the right direction, in terms of providing actionable feedback while also limiting the load on the system in continuing to process an error-prone file.
> > cc: @klinkoberstar @victoriaatraft @reitermb
> 
> potential thought to consider re: how this decision will impact the submission history table. if we stop parsing all together at 500 high priority errors, then when STTs make changes to their files and resubmit to address these errors, it's theoretically possible that the total number of errors on the submission history page goes up on resubmission, not down (due to lower priority errors being evaluated that weren't originally). This may be counterintuitive. if we parse all the records but only show the first 500, the submission history page can still list the true # of records with errors - we will just limit how much detail we show them in the report.
> 
> cc: @ADPennington

I will add in that Kathryn's suggestion makes the changes to accomplish the rejected after 500 priority errors easier to implement. But both options are still relatively easy to implement. I also like the idea that we still aren't obfuscating any of the aggregate errors when doing this. 

Another thought is the report could still include all priority errors to avoid iterative parsing, but still list the file as rejected if it has 500 or more priority errors?

ADPennington commented:
> > > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> > 
> > 
> > > > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> > > 
> > > 
> > > i believe its the former: stop processing altogether when 500 of the most important errors were detected; status would be rejected and error report would include the 500 most important errors detected
> > > our goal is for the STT to review these errors, correct them, and resubmit the file. its likely that this will be an iterative process for the STT since a file with 500 critical errors likely has a lot more than 500 critical errors. the email notification in #3251 will help set the appropriate expectations.
> > > and in case its worth noting for documentation purposes, this strategy is aligned with the legacy system behavior when 500 critical errors were detected (see [system edit code T1-501](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_system_edits_fatal_and_warning.pdf)). the system would return T1-501 as an error but did not include any information about what to fix, so we think this is a step in the right direction, in terms of providing actionable feedback while also limiting the load on the system in continuing to process an error-prone file.
> > > cc: @klinkoberstar @victoriaatraft @reitermb
> > 
> > 
> > potential thought to consider re: how this decision will impact the submission history table. if we stop parsing all together at 500 high priority errors, then when STTs make changes to their files and resubmit to address these errors, it's theoretically possible that the total number of errors on the submission history page goes up on resubmission, not down (due to lower priority errors being evaluated that weren't originally). This may be counterintuitive. if we parse all the records but only show the first 500, the submission history page can still list the true # of records with errors - we will just limit how much detail we show them in the report.
> > cc: @ADPennington
> 
> I will add in that Kathryn's suggestion makes the changes to accomplish the rejected after 500 priority errors easier to implement. But both options are still relatively easy to implement. I also like the idea that we still aren't obfuscating any of the aggregate errors when doing this.
> 
> Another thought is the report could still include all priority errors to avoid iterative parsing, but still list the file as rejected if it has 500 or more priority errors?

how about this? @elipe17 @klinkoberstar 

- reject files that have  >= 500 critical errors
- include all critical errors in the first tab
- include aggregate summary of non-critical errors in 2nd tab

benefits:

- users still have full visibility into all errors, reducing the need to iteratively resubmit
- submission history metadata is aligned with the error report
- critical errors have dedicated worksheet
- workbook size likely to be smaller (since lower priority errors aggregated)

cons:
- list of critical errors could still be overwhelming
- tbd performance impact

elipe17 committed:
- moved prioritized querset func to util to be used in different locations
- moved ParserErrorCategoryChoices to util.py to avoid circular import
- updated get_status to reject file if more than 500 priority errors are generated.

elipe17 committed:
- Updated test to conform to priority errors update

elipe17 committed:
- linting

ADPennington commented:
> > > > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> > > 
> > > 
> > > > > @ADPennington I just want to clarify the following statement: `implement threshold: stop processing file after 500 of the most important errors were detected.`. When we say "stop processing file after 500...", does that mean we want to stop parsing the file altogether if we reach that threshold and mark it rejected? Or, is 500 the max number of priority errors we want to display to STTs at any one time?
> > > > 
> > > > 
> > > > i believe its the former: stop processing altogether when 500 of the most important errors were detected; status would be rejected and error report would include the 500 most important errors detected
> > > > our goal is for the STT to review these errors, correct them, and resubmit the file. its likely that this will be an iterative process for the STT since a file with 500 critical errors likely has a lot more than 500 critical errors. the email notification in #3251 will help set the appropriate expectations.
> > > > and in case its worth noting for documentation purposes, this strategy is aligned with the legacy system behavior when 500 critical errors were detected (see [system edit code T1-501](https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_system_edits_fatal_and_warning.pdf)). the system would return T1-501 as an error but did not include any information about what to fix, so we think this is a step in the right direction, in terms of providing actionable feedback while also limiting the load on the system in continuing to process an error-prone file.
> > > > cc: @klinkoberstar @victoriaatraft @reitermb
> > > 
> > > 
> > > potential thought to consider re: how this decision will impact the submission history table. if we stop parsing all together at 500 high priority errors, then when STTs make changes to their files and resubmit to address these errors, it's theoretically possible that the total number of errors on the submission history page goes up on resubmission, not down (due to lower priority errors being evaluated that weren't originally). This may be counterintuitive. if we parse all the records but only show the first 500, the submission history page can still list the true # of records with errors - we will just limit how much detail we show them in the report.
> > > cc: @ADPennington
> > 
> > 
> > I will add in that Kathryn's suggestion makes the changes to accomplish the rejected after 500 priority errors easier to implement. But both options are still relatively easy to implement. I also like the idea that we still aren't obfuscating any of the aggregate errors when doing this.
> > Another thought is the report could still include all priority errors to avoid iterative parsing, but still list the file as rejected if it has 500 or more priority errors?
> 
> how about this? @elipe17 @klinkoberstar
> 
> * reject files that have  >= 500 critical errors
> * include all critical errors in the first tab
> * include aggregate summary of non-critical errors in 2nd tab
> 
> benefits:
> 
> * users still have full visibility into all errors, reducing the need to iteratively resubmit
> * submission history metadata is aligned with the error report
> * critical errors have dedicated worksheet
> * workbook size likely to be smaller (since lower priority errors aggregated)
> 
> cons:
> 
> * list of critical errors could still be overwhelming
> * tbd performance impact

edit here: after syncing with @elipe17 -- we think its best to go with a status of "partially accepted with errors" since "rejected" files will not yield a summary of errors tab in the report. this also means that we need to make some adjustments to how partially accepted status is defined in the KC cc: @klinkoberstar @victoriaatraft @reitermb

elipe17 committed:
- Updated to make status partially accepted
- Updated xlsx tab names
- Updated tests

elipe17 committed:
- Linting

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

reitermb commented:
@elipe17 is there a sample error report in this new format that we could get for a quick a11y once over? / is there a good sample file that would produce a good error report representing these changes? 

cc @victoriaatraft

elipe17 commented:
> @elipe17 is there a sample error report in this new format that we could get for a quick a11y once over? / is there a good sample file that would produce a good error report representing these changes?
> 
> cc @victoriaatraft

Sure thing, I will attach one here and send it to you both in MM
[2023-Q2-SSP Active Case Data (9).xlsx](https://github.com/user-attachments/files/17821722/2023-Q2-SSP.Active.Case.Data.9.xlsx)

victoriaatraft reviewed (approved):
LGTM

reitermb reviewed (approved):
All good from an a11y perspective! Thanks @elipe17 for the sample report, we'll be using that to write up some follow-on knowledge center adjustments.

raftmsohani reviewed (approved):
LGTM!

ADPennington commented:
> edit here: after syncing with @elipe17 -- we think its best to go with a status of "partially accepted with errors" since "rejected" files will not yield a summary of errors tab in the report. this also means that we need to make some adjustments to how partially accepted status is defined in the KC cc: @klinkoberstar @victoriaatraft @reitermb

@elipe17 + team, after more discussion with @klinkoberstar re: KC content for #3290 , we've decided to hold off on changing the file status to "partially accepted with errors" when > 500 "critical" errors are detected.  The main reason is to limit further confusion. We believe that the current file statuses are trying to proxy multiple indicators of data quality, including: 

- to what extent records in the file are added to the db
- if there are errors detected in the records
- (and now) the severity of the errors in the record

For now, we want this ticket to deliver a restructured view of errors to help STTs focus attention on the errors OFA has identified as critical. ~~We are still in agreement to limit the number of errors in the critical tab of the report to the first 500, with the understanding that the summary tab will capture the frequency of all the errors detected.~~ 


Updated comment on 11/25/2024 to remove cap on critical errors. this will help alleviate the need to resubmit multiple times to fix critical errors. cc: @elipe17 @klinkoberstar

elipe17 committed:
- remove status change based on critical errors

elipe17 committed:
- linting

elipe17 committed:
- fix errro

ADPennington committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+**Date**:     October 20, 2024 <br>
+
+## Summary
+This technical memorandum provides a suggested path to implement a set of new requirements OFA has generated to alleviate the sheer number of parser errors generated during a STT's data submission. OFA has indicated, while still necessary for admins to see, some errors are of a lower priority for STTs to view and even prevent them from using the submission history reports to correct their submiss ...
```
```suggestion
This technical memorandum provides a suggested path to implement a set of new requirements OFA has generated to alleviate the sheer number of parser errors generated during a STT's data submission. OFA has indicated that some errors are of a lower priority for STTs to review and correct. Thus, the OFA team has requested that  "critical" be assigned to parser errors so that the report STTs receive is filtered down to only the critical errors that must be reviewed and fixed. Regardless of how errors are prioritized, STTs will still retain the ability to see a summary of all errors detected in the error report.
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+Currently, error reports are generated in the TDP backend via the `get_xls_serialized_file` function. This function accepts the serialized queryset of the appropriate `ParserError`s queryset. This function the writes an XLSX file and returns it to the user. Apart from the lack of priotization in the report generated from this function, it also introduces the possibility to cause an out of memory  ...
+
+## Out of Scope
+Current requirements from OFA do not require category two errors to be queryable by value and expected value. That feature is out of scope within the tech memo and would require more design and implementation work.
```
@elipe17 is this true? can you remind me what this is in reference to?

elipe17 replied:
For the time being it is true. the OneNote did not specify specific values of the cat2 validators were of interest but that all values of the cat2 validator were of interest. I briefly asked you about it in our MM chat attached below.
![Screenshot 2024-12-05 at 10 12 13 AM](https://github.com/user-attachments/assets/fb4f34a8-d734-4c98-988a-0c8ce23a69ef)

elipe17 replied:
Just for your SA, those messages are from October 18

ADPennington replied:
ahh thank you for clarifying; i misunderstood the statement to mean that we didnt expect any cat 2 errors in the report.

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+Current requirements from OFA do not require category two errors to be queryable by value and expected value. That feature is out of scope within the tech memo and would require more design and implementation work.
+
+## Method/Design
+Given the current OFA requirements, we can implement prioritized errors, and memory efficient report generation without too much work. OFA has provided [this OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) document which outlines the error types, errors, and fields that are most important/prioritized for STTs to see.
```
```suggestion
Given the current OFA requirements, we can implement prioritized/critical errors, and memory efficient report generation without too much work. OFA has provided [this OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) document which outlines the error types, errors, and fields that are most important/prioritized for STTs to see.
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+    return ','.join([i for i in fields_json['friendly_name'].keys()])
+```
+
+### Prioritized Errors
````
```suggestion
### Prioritized/Critical Errors
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+[This OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) is invaluable to the implementation of prioritized errors. Prioritizing errors could be a very large and technically challenging feature involving new migrations, validation/validator refactors, etc... However, this can all be avoided by making a key insight for each  ...
+
+```python
+# All cat2 errors associated with FAMILY_AFFILIATION and (CITIZENSHIP_STATUS or CLOSURE_REASON)
````
🥇 insight

elipe17 replied:
This passage also explains the question you had above wrt querying cat2 validators by value.

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+    ))
+```
+
+The key insight for the category three case is less obvious. Looking at the OneNote, it seems as though we might need to query errors based on field name(s), expected value and actual value. However, for category three errors that information is encoded into the error by its existence. For example, the OneNote indicates that a high priority error a STT should have included in their report is `If  ...
````
pretty accurate, from a technical perspective. For non-technical folks, the critical errors are related to how OFA defines cash assistance receipt, and work-eligible individuals.

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+)
+```
+
+The existence of this error, with these fields, is uniquely defined in the appropriate schemas. The same can be said for the remaining high priority category three errors. Thus, to define the high priority errors we need only know the required field(s) and their error type. Given those pieces of information, queries of the form below can be used to filter STT error reports to only show the highes ...
````
```suggestion
The existence of this error, with these fields, is uniquely defined in the appropriate schemas. The same can be said for the remaining critical category three errors. Thus, to define the high priority errors we need only know the required field(s) and their error type. Given those pieces of information, queries of the form below can be used to filter STT error reports to only show the highest priority errors.
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+              error_type=ParserErrorCategoryChoices.VALUE_CONSISTENCY)
+```
+
+By unioning the category two queries from above with the remainder of the category three queries, a queryset containing only the highest priority errors can be generated and subsequently passed to `get_xls_serialized_file` generate and return the prioritized error report to the requesting STT.
````
```suggestion
By unioning the category two queries from above with the remainder of the category three queries, a queryset containing only the critical errors can be generated and subsequently passed to `get_xls_serialized_file` generate and return the prioritized error report to the requesting STT.
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+- TDP frontend: latency time incurred while generating report
+
+## Use and Test cases to consider
+- Admin can get both prioritized and un-prioritized report
```
```suggestion
- STTs can get both critical and summarized errors in error report
```

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
```diff
@@ -0,0 +1,101 @@
+
+## Use and Test cases to consider
+- Admin can get both prioritized and un-prioritized report
+- STT only recieves prioritized report
```
```suggestion
```

ADPennington commented on tdrs-backend/tdpservice/data_files/util.py:
```diff
@@ -3,54 +3,88 @@
+    PRIORITIZED_CAT3 = (
+        ("FAMILY_AFFILIATION", "SSN"),
+        ("FAMILY_AFFILIATION", "CITIZENSHIP_STATUS"),
+        ("AMT_FOOD_STAMP_ASSISTANCE", "AMT_SUB_CC", "CASH_AMOUNT", "CC_AMOUNT", "TRANSP_AMOUNT"),
```
interesting how this seems to also catch the tribal TANF version of this cat 3 error, which includes the following fields (some of which arent referenced here):

`TANF Child Care Care Amount,Amount of Subsidized Child Care,Cash Amount,Other Amount,Transportation Amount,Amount of Food Stamp Assistance,Transitional Services Amount`

![Screenshot 2024-12-05 094543](https://github.com/user-attachments/assets/329e113e-0135-4d9c-b982-83e595777144)

elipe17 replied:
Yes, the reason that we get both covered is because of the type of query I used. It checks that error contains at least all of the fields listed, but could contain more. That is why we get both since they share the same fields, Tribal just has a few extra.

elipe17 committed:
Update docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
Update docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

Co-authored-by: Alex P.  <63075587+ADPennington@users.noreply.github.com>

elipe17 committed:
- updated based on suggestion

ADPennington commented on docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md:
````diff
@@ -0,0 +1,101 @@
+    return ','.join([i for i in fields_json['friendly_name'].keys()])
+```
+
+### Prioritized/Critical Errors
````
one thing we missed capturing in the onenote is that Aggregate and Stratum files with any 2: Record Value Invalid and 3: Record value consistency errors should be on critical worksheet. 🎫

elipe17 replied:
That is not taken into account in the code either. I will need to add that functionality or we can write another ticket to add it.

ADPennington replied:
#3341

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington commented on tdrs-backend/tdpservice/data_files/util.py:
```diff
@@ -3,54 +3,88 @@
+        ("CITIZENSHIP_STATUS", "WORK_ELIGIBLE_INDICATOR"),
+    )
+
+    # All cat1/4 errors
```
we will need to revisit this in a follow-on 🎫. We noted that we want to exclude the following from STT reports but i think it fell out-of-scope for this ticket when we decided to include the summary worksheet. We'll need to figure out how to handle these. Ultimately, OFA **does not** want these cat 4 errors to prevent records from being stored in the db. Something to explore is whether its easiest to relax the validation or just drop these validators. 

![image](https://github.com/user-attachments/assets/ca8f0ba5-998d-465f-809f-c4fe6becfaec)

ADPennington replied:
#3342

elipe17 committed:
- Comment out sources,list

elipe17 committed:
Merge branch '3155-prioritized-errors' of https://github.com/raft-tech/TANF-app into 3155-prioritized-errors

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors

ADPennington reviewed (approved):
this is in great shape @elipe17 thank you! for team awareness, please note the following @klinkoberstar @victoriaatraft @reitermb :

- errors on the `critical` tab include :
  - all file pre-check errors (aka cat 1)
  - all case consistency errors (aka cat 4) 
  - _some_ record value invalid errors (aka cat 2)
  - _some_ record value consistency errors (aka cat 3)

_I've included a table [here](https://gorafttech.sharepoint.com/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc={c9d41a72-bff8-4eba-b245-bb04ba265e12}&action=edit&wd=target%28Dev%20Notes.one%7C61602a1e-e3d8-4f2e-8744-c4fc299cb56b%2FProposed%20TDP%20Validation%20Strategy%20%28in-progress%5C%29%7Cd222863c-1c4a-4e0e-b87a-d5625455e357%2F%29&wdorigin=NavigationUrl) that captures the fields associated with errors that could appear on this tab, organized by section and error type._ 

- `summary` tab includes aggregates of **all** errors detected by the system
- Error reports for Section 3: Aggregate and Section 4: Stratum files currently do not include any cat 2 or cat 3 errors in the `critical` tab but should. This will be addressed in #3341 
- Some cat 4 errors (for Section 2: Closed files) need to be relaxed/dropped. This will be addressed in #3342

elipe17 committed:
Merge branch 'develop' into 3155-prioritized-errors
